src/workerd/io/trace-stream.c++

```diff
@@ -0,0 +1,937 @@
+  obj.set(js, MESSAGE_STR, js.str(log.message));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::Return& ret, StringCache& cache) {
+  auto obj = js.obj();
+  obj.set(js, TYPE_STR, cache.get(js, RETURN_STR));
+
+  KJ_IF_SOME(info, ret.info) {
+    KJ_SWITCH_ONEOF(info) {
+      KJ_CASE_ONEOF(fetch, tracing::FetchResponseInfo) {
+        obj.set(js, INFO_STR, ToJs(js, fetch, cache));
+      }
+      KJ_CASE_ONEOF(custom, tracing::CustomInfo) {
+        obj.set(js, INFO_STR, ToJs(js, custom.asPtr(), cache));
+      }
+    }
+  }
+
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::Link& link, StringCache& cache) {
+  auto obj = js.obj();
+  obj.set(js, TYPE_STR, cache.get(js, LINK_STR));
+  KJ_IF_SOME(label, link.label) {
+    obj.set(js, LABEL_STR, js.str(label));
+  }
+  obj.set(js, TRACEID_STR, js.str(link.traceId.toGoString()));
+  obj.set(js, INVOCATIONID_STR, js.str(link.invocationId.toGoString()));
+  obj.set(js, SPANID_STR, js.str(link.spanId.toGoString()));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::TailEvent& event, StringCache& cache) {
+  auto obj = js.obj();
+  obj.set(js, TRACEID_STR, js.str(event.traceId.toGoString()));
+  obj.set(js, INVOCATIONID_STR, js.str(event.invocationId.toGoString()));
+  obj.set(js, SPANID_STR, js.str(event.spanId.toGoString()));
+  obj.set(js, TIMESTAMP_STR, js.date(event.timestamp));
+  obj.set(js, SEQUENCE_STR, js.num(event.sequence));
+
+  KJ_SWITCH_ONEOF(event.event) {
+    KJ_CASE_ONEOF(onset, tracing::Onset) {
+      obj.set(js, EVENT_STR, ToJs(js, onset, cache));
+    }
+    KJ_CASE_ONEOF(outcome, tracing::Outcome) {
+      obj.set(js, EVENT_STR, ToJs(js, outcome, cache));
+    }
+    KJ_CASE_ONEOF(hibernate, tracing::Hibernate) {
+      obj.set(js, EVENT_STR, ToJs(js, hibernate, cache));
+    }
+    KJ_CASE_ONEOF(spanOpen, tracing::SpanOpen) {
+      obj.set(js, EVENT_STR, ToJs(js, spanOpen, cache));
+    }
+    KJ_CASE_ONEOF(spanClose, tracing::SpanClose) {
+      obj.set(js, EVENT_STR, ToJs(js, spanClose, cache));
+    }
+    KJ_CASE_ONEOF(mark, tracing::Mark) {
+      KJ_SWITCH_ONEOF(mark) {
+        KJ_CASE_ONEOF(de, tracing::DiagnosticChannelEvent) {
+          obj.set(js, EVENT_STR, ToJs(js, de, cache));
+        }
+        KJ_CASE_ONEOF(ex, tracing::Exception) {
+          obj.set(js, EVENT_STR, ToJs(js, ex, cache));
+        }
+        KJ_CASE_ONEOF(log, tracing::Log) {
+          obj.set(js, EVENT_STR, ToJs(js, log, cache));
+        }
+        KJ_CASE_ONEOF(ret, tracing::Return) {
+          obj.set(js, EVENT_STR, ToJs(js, ret, cache));
+        }
+        KJ_CASE_ONEOF(link, tracing::Link) {
+          obj.set(js, EVENT_STR, ToJs(js, link, cache));
+        }
+        KJ_CASE_ONEOF(attrs, kj::Array<tracing::Attribute>) {
+          obj.set(js, EVENT_STR, ToJs(js, attrs, cache));
+        }
+      }
+    }
+  }
+
+  return obj;
+}
+
+class ServerTopLevelMembrane final: public capnp::MembranePolicy, public kj::Refcounted {
```

Comment:
This is adopted from the class of the same name over is worker-rpc.c++. It can, and probably should be deduplicated into a common utility but will do that separate.